src/main/java/htsjdk/beta/plugin/reads/ReadsBundle.java

```diff
@@ -0,0 +1,170 @@
+
+    /**
+     * @return optional index for this resource, or Optional.empty if not present
+     */
+    public Optional<T> getIndex() {
+        final Supplier<RuntimeException> ex = () -> new RuntimeException("Index resource is present with a null path");
+        final Optional<BundleResource> inputResource = get(BundleResourceType.READS_INDEX);
+        // its OK for there to be no index resource, but if there *is* an index resource, it must contain
+        // a non-null path...
+        return inputResource.isPresent() ?
+                Optional.of((T) inputResource.get().getIOPath().orElseThrow(ex)) :
+                Optional.empty();
+    }
+
+    public static <T extends IOPath> ReadsBundle getReadsBundleFromPath(final T jsonPath) {
+        return new ReadsBundle(IOUtils.getStringFromPath(jsonPath));
+    }
+
+    public static <T extends IOPath> ReadsBundle resolveIndex(T reads) {
+        final Path index = SamFiles.findIndex(reads.toPath());
+        if (index == null) {
+            return new ReadsBundle(reads);
+        }
+        return new ReadsBundle(reads, IOUtils.toHtsPath(index));
+    }
+
+    public static <T extends IOPath> boolean looksLikeAReadsBundle(final T rawReadPath) {
+        return rawReadPath.getURI().getPath().endsWith(BUNDLE_EXTENSION);
+    }
+
+    // try to get the reads for the side effect of validating that the bundle actually contains
+    // a reads resource. throw if its not present
+    private T getReadsResourceOrThrow() {
+        final Supplier<RuntimeException> ex = () -> new RuntimeException("No reads in reads bundle");
+        final Optional<IOPath> ioPath = get(BundleResourceType.READS).orElseThrow(ex).getIOPath();
+        return (T) ioPath.orElseThrow(ex);
+    }
+
+    private static <T extends IOPath> IOPathResource toInputResource(final String providedContentType, final T ioPath) {
+        ValidationUtils.nonNull(ioPath, "ioPath");
+        final Optional<Tuple<String, String>> typePair = getInferredContentTypes(ioPath);
+        if (typePair.isPresent()) {
+            if (providedContentType != null && !typePair.get().a.equals(providedContentType)) {
+                LOG.warn(String.format(
+                        "Provided content type \"%s\" for \"%s\" doesn't match derived content type \"%s\"",
+                        providedContentType,
+                        ioPath.getRawInputString(),
+                        typePair.get().a));
+            }
+            return new IOPathResource(
+                    ioPath,
+                    providedContentType,  // prefer the provided content type
+                    typePair.get().b);
+        } else {
+            return new IOPathResource(
+                    ioPath,
+                    providedContentType);
+        }
+    }
+
+    //try to infer the contentType/subType, i.e., READS/BAM from an IOPath
+    private static <T extends IOPath> Optional<Tuple<String, String>> getInferredContentTypes(final T ioPath) {
```

Comment:
I'm not sure where this sort of guessing code should live but I'd kind of like to separate it from the bundle mechanism itself.  Maybe that's not feasible.

Reply:
I'm not really even sure this is a good idea yet, or whether we need it anywhere. So far I don't use it so I think I'll remove it until we need ti for something.

Reply:
Edit: I take that back - it is being used within this class to sanity check inputs and issue warnings, so I'll leave it for now. I think it makes sense to keep it here since its inferring reads content types.